packages/typescript/internal/ts_project.bzl

```diff
@@ -171,11 +171,15 @@ def _ts_project_impl(ctx):
     # tsc will only produce .json if it also produces .js
     if len(ctx.outputs.js_outs):
         pkg_len = len(ctx.label.package) + 1 if len(ctx.label.package) else 0
-        json_outs = [
-            ctx.actions.declare_file(_join(ctx.attr.out_dir, src.short_path[pkg_len:]))
+        json_file_names = [
```

Comment:
the asymmetry bothers me ... we have an attribute to the rule `js_outs` and the macro that calls the rule determines what output files to expect, so the calculation happens below

`js_outs.extend(_out_paths(srcs, out_dir, root_dir, allow_js, exts))`

can we do the same thing with json_outs - declare that as a new attribute on the rule so the two are handled symmetrically? or is there some reason you weren't able to do that?

Reply:
Hi @alexeagle , thanks for the feedback. I agree with your point here. I think it would actually be best to move the calculation of outputs for the other fields into the rule body. That would fix the issue of not being able to use sources from other rules like `filegroup`, since you need to calculate exact list of the output files. I left a more detailed explanation as a comment on [this issue](https://github.com/bazelbuild/rules_nodejs/issues/2504#issuecomment-912456378).

I'd be happy to update this pull request to include those suggested changes, that would also address your comment here. I might just need some help testing to avoid any regression. Let me know what you think.

Reply:
I'd prefer in this PR to follow the pattern that's already here, and keep that other issue orthogonal. Does it still work for you if you do the calculation in the macro?

Reply:
Unfortunately, it won't work, because I need to resolve the individual input files in order to calculate the output files and this is not possible from within a macro, only from `ctx.files`

Reply:
Let's continue this discussion on the other issue, since the outcome of this appears to be dependent on the output of that: https://github.com/bazelbuild/rules_nodejs/issues/2504

Reply:
why can't the macro predict the output files? can't it determine that from the .json files listed in the srcs like it does for .ts?

again, for the sake of keeping the two orthogonal (this could land now, the other is trickier)

Reply:
the other one landed, time to rebase and get this one done?